Exercise

 1.   Load heart disease dataset in pandas dataframe
 2.   Remove outliers using Z score. Usual guideline is to remove anything that has Z score > 3 formula or Z score < -3
 3.   Convert text columns to numbers using label encoding and one hot encoding
 4.   Apply scaling
 5.   Build a classification model using support vector machine. Use standalone model as well as Bagging model and check if you see any difference in the performance.
 6.   Now use decision tree classifier. Use standalone model as well as Bagging and check if you notice any difference in performance
 7.   Comparing performance of svm and decision tree classifier figure out where it makes most sense to use bagging and why. Use internet to figure out in what conditions bagging works the best.


# Step 0: Import libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pylab as plt
import seaborn as sn

# Step 1: Load Dataset

In [3]:
df = pd.read_csv('/kaggle/input/bagging-exercise/heart.csv')
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [4]:
df.shape

(918, 12)

In [5]:
df.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  918.000000  918.000000   918.000000  918.000000  918.000000   
mean    53.510893  132.396514   198.799564    0.233115  136.809368   
std      9.432617   18.514154   109.384145    0.423046   25.460334   
min     28.000000    0.000000     0.000000    0.000000   60.000000   
25%     47.000000  120.000000   173.250000    0.000000  120.000000   
50%     54.000000  130.000000   223.000000    0.000000  138.000000   
75%     60.000000  140.000000   267.000000    0.000000  156.000000   
max     77.000000  200.000000   603.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  918.000000    918.000000  
mean     0.887364      0.553377  
std      1.066570      0.497414  
min     -2.600000      0.000000  
25%      0.000000      0.000000  
50%      0.600000      1.000000  
75%      1.500000      1.000000  
max      6.200000      1.000000

In [22]:
df.isnull().sum()

Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64

# Step 2: Remove outliers

In [6]:
df1 = df[df.RestingBP>(df.RestingBP.mean()+3*df.RestingBP.std())]
df1 = df[df.RestingBP<(df.RestingBP.mean()+3*df.RestingBP.std())]

In [7]:
df2 = df1[df1.Cholesterol>(df1.Cholesterol.mean()+3*df1.Cholesterol.std())]
df2 = df1[df1.Cholesterol<(df1.Cholesterol.mean()+3*df1.Cholesterol.std())]

In [8]:
df3 = df2[df2.FastingBS>(df2.FastingBS.mean()+3*df2.FastingBS.std())]
df3 = df2[df2.FastingBS<(df2.FastingBS.mean()+3*df2.FastingBS.std())]

In [9]:
df4 = df3[df.MaxHR>(df3.MaxHR.mean()+3*df3.MaxHR.std())]
df4 = df3[df.MaxHR<(df3.MaxHR.mean()+3*df3.MaxHR.std())]

/tmp/ipykernel_42/2676693088.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df4 = df3[df.MaxHR>(df3.MaxHR.mean()+3*df3.MaxHR.std())]
/tmp/ipykernel_42/2676693088.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df4 = df3[df.MaxHR<(df3.MaxHR.mean()+3*df3.MaxHR.std())]


In [10]:
df5 = df4[df.Oldpeak>(df4.Oldpeak.mean()+3*df4.Oldpeak.std())]
df5 = df4[df.Oldpeak<(df4.Oldpeak.mean()+3*df4.Oldpeak.std())]

/tmp/ipykernel_42/2739043989.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df5 = df4[df.Oldpeak>(df4.Oldpeak.mean()+3*df4.Oldpeak.std())]
/tmp/ipykernel_42/2739043989.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  df5 = df4[df.Oldpeak<(df4.Oldpeak.mean()+3*df4.Oldpeak.std())]


In [11]:
df.shape

(918, 12)

In [12]:
df5.shape

(902, 12)

**So we cleared 918 - 902 = 16 outliers!**

# Step 3: Convert text columns to numbers

In [14]:
df5.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [13]:
df5.Sex.unique()

array(['M', 'F'], dtype=object)

In [15]:
df5.ChestPainType.unique()

array(['ATA', 'NAP', 'ASY', 'TA'], dtype=object)

In [16]:
df5.RestingECG.unique()

array(['Normal', 'ST', 'LVH'], dtype=object)

In [17]:
df5.ExerciseAngina.unique()

array(['N', 'Y'], dtype=object)

In [18]:
df5.ST_Slope.unique()

array(['Up', 'Flat', 'Down'], dtype=object)

In [19]:
df6 = df5.copy()

df6.ChestPainType.replace(
    {
        'ATA': 0,
        'NAP': 1,
        'ASY': 2,
        'TA' : 3
    },
    inplace=True
)

df6.RestingECG.replace(
    {
        'Normal': 1,
        'ST': 2,
        'LVH': 3
    },
    inplace=True
)

df6.ExerciseAngina.replace(
    {
        'N': 0,
        'Y': 1
    },
    inplace=True
)

df6.ST_Slope.replace(
    {
        'Down': 1,
        'Flat': 2,
        'Up': 3
    },
    inplace=True
)

df6.head()

Age Sex  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  \
0   40   M              0        140          289          0           1   
1   49   F              1        160          180          0           1   
2   37   M              0        130          283          0           2   
3   48   F              2        138          214          0           1   
4   54   M              1        150          195          0           1   

   MaxHR  ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  
0    172               0      0.0         3             0  
1    156               0      1.0         2             1  
2     98               0      0.0         3             0  
3    108               1      1.5         2             1  
4    122               0      0.0         3             0

In [20]:
df7 = pd.get_dummies(df6, dtype='int', drop_first=True)
df7.head()

Age  ChestPainType  RestingBP  Cholesterol  FastingBS  RestingECG  MaxHR  \
0   40              0        140          289          0           1    172   
1   49              1        160          180          0           1    156   
2   37              0        130          283          0           2     98   
3   48              2        138          214          0           1    108   
4   54              1        150          195          0           1    122   

   ExerciseAngina  Oldpeak  ST_Slope  HeartDisease  Sex_M  
0               0      0.0         3             0      1  
1               0      1.0         2             1      0  
2               0      0.0         3             0      1  
3               1      1.5         2             1      0  
4               0      0.0         3             0      1

# Step 4: Apply Scaling

In [23]:
X = df7.drop('HeartDisease', axis='columns')
y = df7.HeartDisease

In [24]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled

array([[-1.42896269, -1.6990351 ,  0.46089071, ..., -0.84676261,
         1.0456339 ,  0.51485643],
       [-0.47545956, -0.52558207,  1.5925728 , ...,  0.14079864,
        -0.62072967, -1.94228905],
       [-1.74679706, -1.6990351 , -0.10495034, ..., -0.84676261,
         1.0456339 ,  0.51485643],
       ...,
       [ 0.37209878,  0.64787097, -0.10495034, ...,  0.33831089,
        -0.62072967,  0.51485643],
       [ 0.37209878, -1.6990351 , -0.10495034, ..., -0.84676261,
        -0.62072967, -1.94228905],
       [-1.64085227, -0.52558207,  0.3477225 , ..., -0.84676261,
         1.0456339 ,  0.51485643]])

# Step 5: Build a classification model using support vector machine. Use standalone model as well as Bagging model and check if you see any difference in the performance.

In [25]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, stratify=y, random_state=10)

In [26]:
from sklearn import svm
from sklearn.model_selection import cross_val_score

scores = cross_val_score(svm.SVC(), X, y, cv=5)
scores

array([0.62430939, 0.79558011, 0.68888889, 0.72777778, 0.61666667])

In [27]:
scores.mean()

0.6906445672191528

In [29]:
from sklearn.ensemble import BaggingClassifier

bag_model = BaggingClassifier(
    estimator=svm.SVC(), 
    n_estimators=100, 
    max_samples=0.8, 
    oob_score=True,
    random_state=0
)
scores = cross_val_score(bag_model, X, y, cv=5)
scores

array([0.63535912, 0.79558011, 0.68888889, 0.69444444, 0.60555556])

In [30]:
scores.mean()

0.6839656230816453

# Step 6: Now use decision tree classifier. Use standalone model as well as Bagging and check if you notice any difference in performance

In [31]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, stratify=y, random_state=10)

In [34]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score

scores = cross_val_score(DecisionTreeClassifier(), X, y, cv=5)
scores

array([0.76243094, 0.77348066, 0.77222222, 0.78888889, 0.62222222])

In [35]:
scores.mean()

0.7438489871086557

In [36]:
from sklearn.ensemble import BaggingClassifier

bag_model = BaggingClassifier(
    estimator=DecisionTreeClassifier(), 
    n_estimators=100, 
    max_samples=0.8, 
    oob_score=True,
    random_state=0
)
scores = cross_val_score(bag_model, X, y, cv=5)
scores

array([0.87292818, 0.82320442, 0.82222222, 0.81111111, 0.72222222])

In [37]:
scores.mean()

0.8103376304481277

# Step 7: Comparing performance of svm and decision tree classifier

**The best accuracy is with decision tree classifier and bagging (0.81) !**